[TOPIC-BLE-LLCP] Continued work on LLCP

thoh-ot opened the pull request.

thoh-ot committed:
Bluetooth: controller: Host notification node support

Host notifications can either be based on
- LL Data Channel payload or
- A specific rx node payload

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Restructure ull_cp_conn

To smoothen the transition from ull_cp_conn to ll_conn reorder members
in ull_cp_conn to more closely mimic the final placement in ll_conn.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Update LL RX interface

Replace ll_rx_enqueue from llcp module with an external reference, that
is mocked in the unit test.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: bluetooth: llcp: Rewrite tests to use prepare/done flow

Add a flow of prepare/done calls to all tests to better mimic
reality.

Move all Rx handling to done() via a new slist.
Mange test Rx node memory with a memory slab.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: Rename helper enum

Signify that this is only valid for LL Control PDU Payload.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: Divide tests into separate suites

Divide the tests into separate test suites to prepare for future split
of test suites into separate files.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: Rename tests

Rename tests to follow a common pattern:

test_<procedure>_<role>_<initiator>[_<condition>]

where:

<procedure>
name of the procedure (i.e. version_exchange)

<role>
role of the implementation under test:
- mas for master or
- sla for slave.

<initiator>
which device initiates the procedure.
- loc for local side
- rem for remote peer

<condition>
optional; any special condition that test assumes.
ex.: no_ltk, unsupp_feat, limited_memory

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: Refactor test setup logic

Use ztest ztest_unit_test_setup_teardown to reuse the same setup logic
in test cases.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Local PHY Update

Unknown response handling in local PHY Update.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: Add dummy role

Prepare tests for role support.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: PHY_UPDATE_IND decode helper

By similarity to other procedures, add the necessary PDU decode helper.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: PHY Update role support

Add role support to procedure to support both master and slave.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

kruithofa commented on tests/bluetooth/ctrl_llcp/src/main.c:
```diff
@@ -1511,25 +1511,38 @@ void test_api_remote_encryption_start_no_ltk(void)
+			 ztest_unit_test(test_api_disconnect)
+			);
+
+	ztest_test_suite(version_exchange,
```
Eventually we should move these tests to their own test-suite/module/files

thoh-ot replied:
Agree

kruithofa commented on tests/bluetooth/ctrl_llcp/src/main.c:
```diff
@@ -464,7 +464,9 @@ void lt_tx(helper_pdu_opcode_t opcode, struct ull_cp_conn *conn, void *param)
 
 	pdu = (struct pdu_data *) rx->pdu;
-	helper_pdu_encode[opcode](pdu, param);
+	if (helper_pdu_encode[opcode]) {
```
Should we make this into a zassert? If we come here we want to send a PDU, so we should fill in data in this PDU

thoh-ot replied:
I can do that

kruithofa commented on tests/bluetooth/ctrl_llcp/src/main.c:
```diff
@@ -431,11 +431,11 @@ typedef enum {
 	LL_START_ENC_REQ,
 	LL_START_ENC_RSP,
-} helper_opcode_t;
+} helper_pdu_opcode_t;
```
Note extra comma at end of previous line

thoh-ot replied:
Common construct found all over the code base

thoh-ot committed:
tests: Bluetooth: llcp: Check for valid helper

Check that there is valid helper function before calling it.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: PHY Update collision resolving

Add master and slave collision resolving and test.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Fix Request FSM

As calling lr_act_run()/rr_act_run() might lead to an internal state
change, do not overwrite the state from the outside.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Refactor state setting for LR/RR FSM

Use setter function and enum for Local/Remote Reqest FSM.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: Better assert messages

Add descriptive error messages and call site information.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: llcp: PHY Update resume pending proc

When a remote procedure is active and a local procedure is initiated,
the local procedure is paused and waits for the remote procedure to
finish.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Resume PHY Update procedure

Resume PHY Update procedure that is either waiting for Tx buffer or
was paused due to collision resolving.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Remote PHY Update

Add Local PHY Update procedure state machine and test.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Local PHY Update

Add Local PHY Update procedure state machine and test.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
Bluetooth: controller: Add collision resolving interface

Collision resolved between active Local Procedure FSM and
Remote Request FSM.

Remote side:
* Only non-colliding procedure executed.
* Acts on the incompatibility signal.
* Outputs collision signal.

Local side:
* Pending procedure always executed.
* Outputs incompatibility signal.
* Acts on collision signal.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

kruithofa commented on tests/bluetooth/ctrl_llcp/src/main.c:
```diff
@@ -510,6 +515,25 @@ void ut_rx_q_is_empty()
 	zassert_is_null(ntf, NULL);
 }
 
+void prepare(struct ull_cp_conn *conn)
```
Rename to something more descriptive, like event_prepare
